### PR TITLE
[Keyboard] Maartenwut Wonderland PCB

### PR DESCRIPTION
**Pull request by yanfali**
Adds Maartenwut Wonderland PCB to QMK. Will be used in the upcoming lubrigante GB and sold separately with a mini-usb connector for people who want to replace their Alice PCBs.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/wonderland/config.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#define MATRIX_ROW_PINS {B0,D1,D2,D3,D5}
+#define MATRIX_COL_PINS {F0,F1,F4,F5,F6,F7,E6,C7,C6,B6,B5,B4,D7,D6,D4}
+#define UNUSED_PINS
+
+
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCING 5
```
Comment (fauxpark) [inline keyboards/wonderland/config.h]:
```suggestion
#define DEBOUNCE 5
```

File: keyboards/wonderland/wonderland.c [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	matrix_scan_user();
+};
+
+void led_init_ports(void) {
+    // * Set our LED pins as output
+    DDRB |= (1 << 1) | (1 << 2) | (1 << 3);
+}
+
+void led_set_kb(uint8_t usb_led) {
+	if (usb_led & (1 << USB_LED_NUM_LOCK)) {
```
Comment (fauxpark) [inline keyboards/wonderland/wonderland.c]:
```suggestion
	if (IS_LED_ON(usb_led, USB_LED_NUM_LOCK)) {
```

File: keyboards/wonderland/wonderland.c [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        DDRB |= (1 << 1);
+    } else {
+        DDRB &= ~(1 << 1);
+    }
+
+    if (usb_led & (1 << USB_LED_CAPS_LOCK)) {
```
Comment (fauxpark) [inline keyboards/wonderland/wonderland.c]:
```suggestion
    if (IS_LED_ON(usb_led, USB_LED_CAPS_LOCK)) {
```

File: keyboards/wonderland/wonderland.c [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        DDRB |= (1 << 2);
+    } else {
+        DDRB &= ~(1 << 2);
+    }
+
+    if (usb_led & (1 << USB_LED_SCROLL_LOCK)) {
```
Comment (fauxpark) [inline keyboards/wonderland/wonderland.c]:
```suggestion
    if (IS_LED_ON(usb_led, USB_LED_SCROLL_LOCK)) {
```

File: keyboards/wonderland/wonderland.c [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+};
+
+void led_init_ports(void) {
+    // * Set our LED pins as output
+    DDRB |= (1 << 1) | (1 << 2) | (1 << 3);
+}
+
+void led_set_kb(uint8_t usb_led) {
+	if (usb_led & (1 << USB_LED_NUM_LOCK)) {
+        DDRB |= (1 << 1);
```
Comment (fauxpark) [inline keyboards/wonderland/wonderland.c]:
```suggestion
        setPinOutput(B1);
```

File: keyboards/wonderland/wonderland.c [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    } else {
+        DDRB &= ~(1 << 1);
```
Comment (fauxpark) [inline keyboards/wonderland/wonderland.c]:
```suggestion
        setPinInput(B1);
```

File: keyboards/wonderland/wonderland.c [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    if (usb_led & (1 << USB_LED_CAPS_LOCK)) {
+        DDRB |= (1 << 2);
```
Comment (fauxpark) [inline keyboards/wonderland/wonderland.c]:
```suggestion
        setPinOutput(B2);
```

File: keyboards/wonderland/wonderland.c [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    } else {
+        DDRB &= ~(1 << 2);
```
Comment (fauxpark) [inline keyboards/wonderland/wonderland.c]:
```suggestion
        setPinInput(B2);
```

File: keyboards/wonderland/wonderland.c [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    if (usb_led & (1 << USB_LED_SCROLL_LOCK)) {
+        DDRB |= (1 << 3);
```
Comment (fauxpark) [inline keyboards/wonderland/wonderland.c]:
```suggestion
        setPinOutput(B3);
```

File: keyboards/wonderland/wonderland.c [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    } else {
+        DDRB &= ~(1 << 3);
```
Comment (fauxpark) [inline keyboards/wonderland/wonderland.c]:
```suggestion
        setPinInput(B3);
```

File: keyboards/wonderland/wonderland.c [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+void matrix_scan_kb(void) {
+	// put your looping keyboard code here
+	// runs every cycle (a lot)
+	matrix_scan_user();
+};
+
+void led_init_ports(void) {
+    // * Set our LED pins as output
+    DDRB |= (1 << 1) | (1 << 2) | (1 << 3);
```
Comment (fauxpark) [inline keyboards/wonderland/wonderland.c]:
```suggestion
    setPinOutput(B1);
    setPinOutput(B2);
    setPinOutput(B3);
```

File: keyboards/wonderland/wonderland.c (hunk 5 above)
Comment (fauxpark) [reply keyboards/wonderland/wonderland.c]: Actually, should these not be `writePinHigh`/`writePinLow`?

Comment (yanfali) [reply keyboards/wonderland/wonderland.c]: Is there a doc for new convention?

Comment (yanfali) [reply keyboards/wonderland/wonderland.c]: Looks like it should writePinLow and writePinHigh, will fix. https://docs.qmk.fm/#/custom_quantum_functions?id=example-led_set_user-implementation

Comment (fauxpark) [reply keyboards/wonderland/wonderland.c]: Not sure if I'd call it documentation, but... https://docs.qmk.fm/#/custom_quantum_functions?id=example-led_set_user-implementation

File: keyboards/wonderland/wonderland.c (hunk 10 above)
Comment (yanfali) [reply keyboards/wonderland/wonderland.c]: thank you for hint

File: keyboards/wonderland/wonderland.c (hunk 9 above)
Comment (yanfali) [reply keyboards/wonderland/wonderland.c]: thanks for hint

File: keyboards/wonderland/wonderland.c (hunk 5 above)
Comment (fauxpark) [reply keyboards/wonderland/wonderland.c]: The macros to use depend on how the LED is wired; that example assumes the pins are initialised to output (which is the case with this board).

Though, I am not sure if initing to input is electrically valid for such a circuit. Maybe switching between tristate and input low can toggle the LED in current sink configuration...

**Comment by yanfali** [comment]
@fauxpark good points. I will get keebs to test firmware before adding to main repo.

File: keyboards/wonderland/rules.mk [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+COMMAND_ENABLE = yes    # Commands for debug and configuration
+NKRO_ENABLE = no		# USB Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+RGBLIGHT_ENABLE = yes   # Enable keyboard underlight functionality (+4870)
+BACKLIGHT_ENABLE = no  # Enable keyboard backlight functionality (+1150)
+MIDI_ENABLE = no 		# MIDI controls
+AUDIO_ENABLE = no
+UNICODE_ENABLE = no 		# Unicode
+BLUETOOTH_ENABLE = no # Enable Bluetooth with the Adafruit EZ-Key HID
+AUTO_SHIFT_ENABLE = no
+VELOCIKEY_ENABLE = no
```
Comment (noroadsleft) [inline keyboards/wonderland/rules.mk]: You have `VLK_TOG` in the default keymap (on Line 19) but Velocikey is turned off.

File: keyboards/wonderland/config.h [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0xA71C
+#define DEVICE_VER      0x0003
+#define MANUFACTURER    Maartenwut
+#define PRODUCT         Wonderland
+#define DESCRIPTION     "QMK Firmware for the Wonderland, a TGR Alice compatible PCB"
+
+/* key matrix size */
+#define MATRIX_ROWS 5
+#define MATRIX_COLS 16
```
Comment (noroadsleft) [inline keyboards/wonderland/config.h]: You have 16 columns set here but `MATRIX_COL_PINS` only has 15 pins defined.

File: keyboards/maartenwut/wonderland/keymaps/keebs/keymap.c [hunk 14]
```diff
@@ -0,0 +1,23 @@
+#include QMK_KEYBOARD_H
+
+#define ___ KC_NO
```
Comment (noroadsleft) [inline keyboards/maartenwut/wonderland/keymaps/keebs/keymap.c]: I would much prefer if the QMK-native seven-character alias was used for this.

```suggestion
```

File: keyboards/maartenwut/wonderland/config.h [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0xA71C
+#define DEVICE_VER      0x0003
+#define MANUFACTURER    Maartenwut
+#define PRODUCT         Wonderland
+#define DESCRIPTION     "QMK Firmware for the Wonderland"
```
Comment (adrientetar) [inline keyboards/maartenwut/wonderland/config.h]: Afaict, this is what will show when plugging in the keyboard. I would rather name it something like "Wonderland Keyboard" or put keyboard in the name.

Comment (fauxpark) [reply keyboards/maartenwut/wonderland/config.h]: `DESCRIPTION` is actually only compiled in if you have Command enabled, and it's not used as part of the device name.

This is the only reference to it in the entire repo:
https://github.com/qmk/qmk_firmware/blob/c0e6734d3fc7f777d60ad7bb31002521e9a1ee26/tmk_core/common/command.c#L176-L183
It *was* previously used as part of the device name for Adafruit Bluetooth boards, but for the exact reason you state, it was removed a while back (#5028).

Comment (yanfali) [reply keyboards/maartenwut/wonderland/config.h]: This is what the vendor asked for, so I'm not going to change it without their approval. As fauxpark pointed out, it's only used in the COMMAND version feature.

**Comment by yanfali** [comment]
TODO: info.json and README.md

**Comment by yanfali** [comment]
@noroadsleft you still have issues with this PR?

**Pull request merged**